tests/vertexonly/test_vertex_only_fs.py

```diff
@@ -0,0 +1,138 @@
+from firedrake import *
+import pytest
+import numpy as np
+from mpi4py import MPI
+
+
+# Utility Functions
+
+@pytest.fixture(params=[pytest.param("interval", marks=pytest.mark.xfail(reason="swarm not implemented in 1d")),
+                        "square",
+                        pytest.param("extruded", marks=pytest.mark.xfail(reason="extruded meshes not supported")),
+                        "cube",
+                        "tetrahedron",
+                        pytest.param("immersedsphere", marks=pytest.mark.xfail(reason="immersed parent meshes not supported")),
+                        pytest.param("periodicrectangle", marks=pytest.mark.xfail(reason="meshes made from coordinate fields are not supported"))])
+def parentmesh(request):
+    if request.param == "interval":
+        return UnitIntervalMesh(1)
+    elif request.param == "square":
+        return UnitSquareMesh(1, 1)
+    elif request.param == "extruded":
+        return ExtrudedMesh(UnitSquareMesh(1, 1), 1)
+    elif request.param == "cube":
+        return UnitCubeMesh(1, 1, 1)
+    elif request.param == "tetrahedron":
+        return UnitTetrahedronMesh()
+    elif request.param == "immersedsphere":
+        return UnitIcosahedralSphereMesh()
+    elif request.param == "periodicrectangle":
+        return PeriodicRectangleMesh(3, 3, 1, 1)
+
+
+@pytest.fixture(params=[0, 1, 100], ids=lambda x: f"{x}-coords")
+def vertexcoords(request, parentmesh):
+    size = (request.param, parentmesh.geometric_dimension())
+    return pseudo_random_coords(size)
+
+
+def pseudo_random_coords(size):
+    """
+    Get an array of pseudo random coordinates with coordinate elements
+    between -0.5 and 1.5. The random numbers are consistent for any
+    given `size` since `numpy.random.seed(0)` is called each time this
+    is used.
+    """
+    np.random.seed(0)
+    a, b = -0.5, 1.5
+    return (b - a) * np.random.random_sample(size=size) + a
+
+
+# Function Space Generation Tests
+
+def functionspace_tests(vm, family, degree):
+    # Prep: Get number of cells
+    num_cells_mpi_global = MPI.COMM_WORLD.allreduce(vm.num_cells(), op=MPI.SUM)
+    # Can create function space
+    V = FunctionSpace(vm, family, degree)
+    # Can create function on function spaces
+    f = Function(V)
+    g = Function(V)
+    # Can interpolate and Galerkin project onto functions
+    gdim = vm.geometric_dimension()
+    if gdim == 1:
+        x, = SpatialCoordinate(vm)
+        f.interpolate(x)
+        g.project(x)
+    elif gdim == 2:
+        x, y = SpatialCoordinate(vm)
+        f.interpolate(x*y)
+        g.project(x*y)
+    elif gdim == 3:
+        x, y, z = SpatialCoordinate(vm)
+        f.interpolate(x*y*z)
+        g.project(x*y*z)
+    # Get exact values at coordinates with maintained ordering
+    assert np.shape(f.dat.data_ro)[0] == np.shape(vm.coordinates.dat.data_ro)[0]
```

Comment:
Prefer property accessors `f.dat.data_ro.shape`. Really you're checking that the `node_set` of the two function spaces is the same here, why not do that?

Reply:
> Really you're checking that the node_set of the two function spaces is the same here, why not do that?

I'm afraid I don't quite understand. I want to know that the data in the function _is_ the coordinates field of the mesh. How does comparing the `node_set` tell me that?

Reply:
> I'm afraid I don't quite understand. I want to know that the data in the function is the coordinates field of the mesh. How does comparing the node_set tell me that?

```python
import numpy
assert numpy.shape(numpy.arange(20).reshape(10, 2))[0]  == numpy.shape(numpy.empty(200).reshape(10, 20))[0]
```

I do not think this test is doing what you think it is doing.

Reply:
Correction - the test is that we get the coordinate field with the correct expression applied (i.e. x in 1D, x*y in 2D, x*y*z in 3D)

Reply:
Perhaps it would help to say that the comment covers this assert and the one below it? The first thing I do is check the `dat` has the number of entries I expect, then I check the values are correct.

Reply:
Why not just `assert numpy.allclose(f.dat.data_ro, numpy.prod(vm.coordinates.dat.data_ro, axis=1))` then?

FWIW, I would argue for fewer comments that describe _what_ the code is doing. One should aim to write code that makes sense without comments and then use them sparingly. Mostly to describe _why_ we're doing something

Reply:
I agree. Changed.

> FWIW, I would argue for fewer comments that describe what the code is doing. One should aim to write code that makes sense without comments and then use them sparingly. Mostly to describe why we're doing something

A balance I always try to strike - though I have a habit of writing a comment saying what I want to do before I do it to keep track! It's hard sometimes to know what is obvious and what isn't without being a more experienced developer in a given codebase (and perhaps knowing what's _pythonic_ a little better than I do - I wrote my first python code about 8 months ago).

Reply:
So I think that's a reasonable approach, but I think it's worthwhile, having filled in the idea with implementation, going back and refactoring the comments too. This is easier if you do initial prototyping development "out of tree": if what you need to do is introduce some whole new functionality, it can be implemented as an "example" and then only later once you're reasonably happy with the design bring it in and fit it in the right place.

Reply:
Thanks. I'll bare that in mind going ahead. Would be keen to learn more about your development workflow if you have any other tips. Will resolve this comment for now.